source/common/http/async_client_impl.h

```diff
@@ -411,15 +453,27 @@ class AsyncStreamImpl : public AsyncClient::Stream,
   void encodeTrailers(ResponseTrailerMapPtr&& trailers) override;
   ResponseTrailerMapOptRef responseTrailers() const override { return {}; }
   void encodeMetadata(MetadataMapPtr&&) override {}
-  void onDecoderFilterAboveWriteBufferHighWatermark() override { ++high_watermark_calls_; }
+  void onDecoderFilterAboveWriteBufferHighWatermark() override {
+    ++high_watermark_calls_;
+    if (watermark_callbacks_.has_value()) {
+      watermark_callbacks_->get().onDecoderFilterAboveWriteBufferHighWatermark();
+    }
+  }
   void onDecoderFilterBelowWriteBufferLowWatermark() override {
     ASSERT(high_watermark_calls_ != 0);
     --high_watermark_calls_;
+    if (watermark_callbacks_.has_value()) {
+      watermark_callbacks_->get().onDecoderFilterBelowWriteBufferLowWatermark();
+    }
   }
   void addDownstreamWatermarkCallbacks(DownstreamWatermarkCallbacks&) override {}
   void removeDownstreamWatermarkCallbacks(DownstreamWatermarkCallbacks&) override {}
-  void setDecoderBufferLimit(uint32_t) override {}
-  uint32_t decoderBufferLimit() override { return 0; }
+  void setDecoderBufferLimit(uint32_t limit) override {
```

Comment:
The goal here is to have exactly the same behavior as before if no buffer limit was specified in options.

Reply:
We could always stream infinite data via the AsyncStream API -- just keep calling `sendData`. I could set the default to be kBufferLimitForRetry (currently 64KiB). Another option (which I maybe like more) is I could pipe through the `retry_shadow_buffer_limit_` from the main Router filter to here instead of using the downstream filter chain `decoderBufferLimit`.

Reply:
but previously didn't we only do shadow streams if we were willing to buffer the whole body?  So there was an implicit limit where now there may not be?

Reply:
right; my question is "What should that limit be?".

Reply:
Is the question here just around what happens if setDecoderBufferLimit is called more than once? I would think it should override the previous value if we expect multiple calls, or ASSERT if we don't expect multiple calls

Reply:
Made it an ENVOY_BUG. The only place that this could be called is from the router filter, and there's no calls in router.cc.